runtime/lua/vim/treesitter/_fold.lua

```diff
@@ -147,6 +147,23 @@ local function normalise_erow(bufnr, erow)
   return math.min(erow or max_erow, max_erow)
 end
 
+---@type table<string, Query|false|nil>
+--- `false` if query was not found
+local queries = {}
+
+-- TODO: use generalized memoize / LRU. #23233
+---@return Query|nil
+local function get_query(lang)
+  local query = queries[lang]
+  if query == false then
+    return nil
+  elseif query == nil then
+    query = ts.query.get(lang, 'folds')
```

Comment:
I thought ts.query.get is already cached?

Reply:
No, it isn't. 
https://github.com/neovim/neovim/blob/54807231c03495d6a322885d3ae7061e47cac490/runtime/lua/vim/treesitter/query.lua#L210-L221

`explicit_queries` only stores queries set by `vim.treesitter.query.set()`, and `get_files` and `read_query_files` don't do caching either.

Reply:
Ah we cache the parsing (https://github.com/neovim/neovim/blob/54807231c03495d6a322885d3ae7061e47cac490/runtime/lua/vim/treesitter/query.lua#L251) but not the file reading.

Would it be possible to add the same parsing caching to file reading via a weak table?

Reply:
Added